app/gui2/src/providers/portInfo.ts

```diff
@@ -1,13 +1,13 @@
 import { createContextStore } from '@/providers'
+import type { AstId, TokenId } from '@/util/ast/abstract'
 import { identity } from '@vueuse/core'
-import type { ExprId } from 'shared/yjsModel'
 
 declare const portIdBrand: unique symbol
 /**
  * Port identification. A port represents a fragment of code displayed/modified by the widget;
  * usually Ast nodes, but other ids are also possible (like argument placeholders).
  */
-export type PortId = ExprId | (string & { [portIdBrand]: never })
+export type PortId = AstId | TokenId | (string & { [portIdBrand]: never })
```

Comment:
I wonder if we actually expect `TokenId` here? The connected/connectable ports should be always AST, possibly identifier.

Reply:
The input to a widget is not the same as other `PortId`s--it can be either a non-token `PortId` or a `TokenId`. This distinction was hidden in the past because we used `ExprId`s, which could refer to an Ast or a Token. I've restricted the definition of `PortId` to exclude tokens. Widgets that need to use their input as a port now cast it to `AsNot<TokenId>`. I don't think this is a perfect solution but it makes explicit an assumption we have been relying on, and it is now the responsibility of each widget to know that its input can't be a bare token.